### PR TITLE
Feat rag add more features

### PR DESCRIPTION
**Pull request by better629**
**Features**
<!-- Clear and direct description of the submit features. -->
<!-- If it's a bug fix, please also paste the issue link. -->

- rag add engine like es
- rag add rerank like flare, object-rerank-by-field-val, colbert
    
**Feature Docs**
<!-- The RFC, tutorial, or use cases about the feature if it's a pretty big update. If not, there is no need to fill. -->

**Influence**
<!-- Tell me the impact of the new feature and I'll focus on it.  -->

**Result**
<!-- The screenshot/log of unittest/running result -->

```
========================================================================================================= test session starts ==========================================================================================================
collected 29 items                                                                                                                                                                                                                     

tests/metagpt/rag/engines/test_simple.py .....                                                                                                                                                                                   [ 17%]
tests/metagpt/rag/factories/test_base.py ..........                                                                                                                                                                              [ 51%]
tests/metagpt/rag/factories/test_ranker.py .....                                                                                                                                                                                 [ 68%]
tests/metagpt/rag/factories/test_retriever.py ......                                                                                                                                                                             [ 89%]
tests/metagpt/rag/retrievers/test_bm25_retriever.py .                                                                                                                                                                            [ 93%]
tests/metagpt/rag/retrievers/test_faiss_retriever.py .                                                                                                                                                                           [ 96%]
tests/metagpt/rag/retrievers/test_hybrid_retriever.py .                                                                                                                                                                          [100%]

=========================================================================================================== warnings summary ===========================================================================================================
=================================================================================================== 29 passed, 7 warnings in 10.90s ====================================================================================================
```

**Other**
<!-- Something else about this PR. -->

### CODE REVIEW

File: examples/rag_pipeline.py [hunk 1]
```diff
@@ -24,6 +28,17 @@
...
 
+def catch_exception(func):
+    @wraps(func)
+    async def wrapper(*args, **kwargs):
+        try:
+            return await func(*args, **kwargs)
+        except Exception as e:
+            logger.error(f"{func.__name__} exception: {e}")
+
+    return wrapper
```
Comment (geekan) [inline examples/rag_pipeline.py:39]: use handle_exception

File: examples/rag_pipeline.py [hunk 2]
```diff
@@ -154,20 +180,43 @@ async def init_and_query_chromadb(self):
         """
         self._print_title("Init And Query ChromaDB")
 
-        # save index
+        # 1.save index
```
Comment (geekan) [inline examples/rag_pipeline.py:183]: 1. a
2. b
3. c

- maintain space after "1."

File: requirements.txt [hunk 3]
```diff
@@ -10,6 +10,16 @@ typer==0.9.0
...
+llama-index-core==0.10.15
+llama-index-embeddings-azure-openai==0.1.6
+llama-index-embeddings-openai==0.1.5
+llama-index-llms-azure-openai==0.1.4
+llama-index-readers-file==0.1.4
+llama-index-retrievers-bm25==0.1.3
+llama-index-vector-stores-faiss==0.1.1
+llama-index-vector-stores-elasticsearch==0.1.6
+llama-index-postprocessor-colbert-rerank==0.1.1
+chromadb==0.4.23
```
Comment (geekan) [inline requirements.txt:22]: update in setup.py would be better.

File: examples/rag_pipeline.py (hunk 1 above)
Comment (seehi) [reply examples/rag_pipeline.py]: fixed

File: examples/rag_pipeline.py (hunk 2 above)
Comment (seehi) [reply examples/rag_pipeline.py]: fixed

File: requirements.txt (hunk 3 above)
Comment (seehi) [reply requirements.txt]: fixed

**Review by geekan — APPROVED** [review]
LGTM

**Pull request merged**